components/scream/src/physics/shoc/tests/shoc_run_and_cmp.cpp

```diff
@@ -0,0 +1,200 @@
+                    const Scalar* b, const Int& n, const Real& tol) {
+  Int nerr = 0;
+  Real den = 0;
+  for (Int i = 0; i < n; ++i)
+    den = std::max(den, std::abs(a[i]));
+  Real worst = 0;
+  for (Int i = 0; i < n; ++i) {
+    if (std::isnan(a[i]) || std::isinf(a[i]) ||
+        std::isnan(b[i]) || std::isinf(b[i])) {
+      ++nerr;
+      continue;
+    }
+    const auto num = std::abs(a[i] - b[i]);
+    if (num > tol*den) {
+      ++nerr;
+      worst = std::max(worst, num);
+    }
+  }
+  if (nerr) {
+    std::cout << label << " nerr " << nerr << " worst " << (worst/den)
+              << " with denominator " << den << "\n";
+  }
+  return nerr;
+}
+
+Int compare (const std::string& label, const double& tol,
+             const FortranData::Ptr& ref, const FortranData::Ptr& d) {
+  Int nerr = 0;
+  FortranDataIterator refi(ref), di(d);
+  scream_assert(refi.nfield() == di.nfield());
+  for (Int i = 0, n = refi.nfield(); i < n; ++i) {
+    const auto& fr = refi.getfield(i);
+    const auto& fd = di.getfield(i);
+    scream_assert(fr.size == fd.size);
+    nerr += compare(label + std::string(" ") + fr.name,
+                    fr.data, fd.data, fr.size, tol);
+  }
+  return nerr;
+}
+
+struct Baseline {
+  Baseline () {
+    params_.push_back({1, 128, 300});
+  }
+
+  Int generate_baseline (const std::string& filename, bool use_fortran) {
+    auto fid = FILEPtr(fopen(filename.c_str(), "w"));
+    scream_require_msg( fid, "generate_baseline can't write " << filename);
+    Int nerr = 0;
+    for (auto ps : params_) {
+      // Run reference shoc on this set of parameters.
+      const auto d = ic::Factory::create(ps.shcol, ps.nlev);
+      set_params(ps, *d);
+      shoc_init(ps.nlev, use_fortran);
+      shoc_main(*d);
+      write(fid, d);
+
+      // Save the fields to the baseline file.
+    }
+    return nerr;
+  }
+
+  Int run_and_cmp (const std::string& filename, const double& tol, bool use_fortran) {
+    auto fid = FILEPtr(fopen(filename.c_str(), "r"));
+    scream_require_msg( fid, "generate_baseline can't read " << filename);
+    Int nerr = 0, ne;
+    int case_num = 0;
+    for (auto ps : params_) {
+      case_num++;
+      // Read the reference impl's data from the baseline file.
+      const auto d_ref = ic::Factory::create(ps.shcol, ps.nlev);
+      set_params(ps, *d_ref);
+      // Now run a sequence of other impls. This includes the reference
+      // implementation b/c it's likely we'll want to change it as we go.
+      {
+        const auto d = ic::Factory::create(ps.shcol, ps.nlev);
+        set_params(ps, *d);
+        shoc_init(ps.nlev, use_fortran);
+        std::cout << "--- checking case # " << case_num << " ---\n" << std::flush;
+        read(fid, d_ref);
+        shoc_main(*d);
+        ne = compare("ref", tol, d_ref, d);
+        if (ne) std::cout << "Ref impl failed.\n";
+        nerr += ne;
+      }
+    }
+    return nerr;
+  }
+
+private:
+
+  struct ParamSet {
+    Int shcol, nlev;
+    Real dtime;
```

Comment:
You'll need `Int nadv`, which needs to be consistent with `nlev` and `dtime` for CFL reasons.